src/librustc_mir/transform/instrument_coverage.rs

```diff
@@ -136,7 +149,7 @@ impl<'tcx> Instrumentor<'tcx> {
             new_block,
             TerminatorKind::Call {
                 func: count_code_region_fn,
-                args: vec![counter_index],
+                args,
                 // new_block will swapped with the next_block, after applying patch
                 destination: Some((Place::from(temp), new_block)),
```

Comment:
@wesleywiser @bjorn3 - I don't (strictly speaking) have to address this in the current PR, but when I start injecting the calls to `coverage_counter_add`, `coverage_counter_subtract`, etc., I expect to also create a `temp` for the returned "unit", with associated `StorageLive` (in the Call `BasicBlock`) and `StorageDead` in the destination `BasicBlock`.

And if the backend implementation for these new intrinsics does not result in an LLVM intrinsic call (as is the case, per `rustc_codegen_llvm::is_codegen_intrinsic()`), the `StorageLive` and `StorageDead` calls and associated temp value have no purpose.

I don't want to make that decision during MIR generation. Whether the Call is converted into a backend call or not is (in my opinion) a codegen-time decision.

Will these unused `StorageLive` and `StorageDead` calls and the unit temp be optimized out by LLVM automatically? Or do I need to do something during codegen to effectively "erase" the storage statements from the LLVM IR in this case?

I see there are some intrinsics in `rustc_codegen_ssa::block` and `rustc_codegen_llvm::intrinsic` that also skip LLVM IR call generation, so maybe this is a problem/pattern that has already been solved; but they are not all the same.

Any pointers?

Thanks!

Reply:
`temp` is of type `()`, which isn't stored on the stack. This makes `StorageLive` and `StorageDead` a nop during codegen. `StorageLive` and `StorageDead` were originally introduced to tell LLVM when stack slots were live or dead. (They are now also used by MIR borrowck and MIRI)

Reply:
Cool! Sounds like I don't need to worry about it, then. Thanks!

Reply:
The mir-opt working group has also discussed optimizing out `StorageLive` and `StorageDead` for ZSTs like `()` before. I wouldn't worry about the performance of this yet since we can optimize this later if need be.